source/common/upstream/original_dst_cluster.cc

```diff
@@ -0,0 +1,163 @@
+      auto& localAddr = connection->localAddress();
+
+      // Check if a host with the destination address is already in the host set.
+      auto it = host_map_.find(localAddr.asString());
+
+      if (it != host_map_.end()) {
+        // Use the existing host
+        auto host = it->second;
+        ENVOY_LOG(debug, "FOUND HOST {}", host->address()->asString());
+        host->weight(2); // Mark as used.
+        return host;
+      } else {
+        // Add a new host
+        auto localIP = localAddr.ip();
+        if (localIP) {
+          Network::Address::InstanceConstSharedPtr hostIpPort;
+          if (localIP->version() == Network::Address::IpVersion::v4) {
+            hostIpPort.reset(
+                new Network::Address::Ipv4Instance(localIP->addressAsString(), localIP->port()));
+          } else {
+            hostIpPort.reset(
+                new Network::Address::Ipv6Instance(localIP->addressAsString(), localIP->port()));
+          }
+          HostSharedPtr host = parent_.createHost(hostIpPort);
+
+          ENVOY_LOG(debug, "CREATED HOST {}", host->address()->asString());
+          // Add the new host to the map.  We just failed to find it in
+          // our local map above, so just insert it.
+          host_map_.insert({localAddr.asString(), host});
+          host->weight(2); // Mark as used.
+          return host;
+        } else {
+          ENVOY_LOG(debug, "FAILED TO CREATE HOST FOR {}", localAddr.asString());
+        }
+      }
+    }
+  }
+
+  ENVOY_LOG(warn, "original_dst_load_balancer: No downstream connection or no original_dst.");
+  return nullptr;
+}
+
+OriginalDstCluster::OriginalDstCluster(const Json::Object& config, Runtime::Loader& runtime,
+                                       Stats::Store& stats,
+                                       Ssl::ContextManager& ssl_context_manager,
+                                       Event::Dispatcher& dispatcher, bool added_via_api)
+    : ClusterImplBase(config, runtime, stats, ssl_context_manager, added_via_api),
+      dispatcher_(dispatcher), cleanup_interval_ms_(std::chrono::milliseconds(
+                                   config.getInteger("cleanup_interval_ms", 5000))),
+      cleanup_timer_(dispatcher.createTimer([this]() -> void { cleanup(); })) {
+  size_t n_hosts = 0;
+  try {
+    std::vector<Json::ObjectSharedPtr> hosts_json = config.getObjectArray("hosts");
+    n_hosts = hosts_json.size();
+  } catch (...) {
+  } // Catch missing hosts config.
+  if (n_hosts != 0) {
+    throw EnvoyException("original_dst clusters must have no hosts configured");
+  }
+
+  cleanup_timer_->enableTimer(cleanup_interval_ms_);
+}
+
+// Called by a worker thread.
+HostSharedPtr OriginalDstCluster::createHost(Network::Address::InstanceConstSharedPtr address) {
+  HostSharedPtr host(
+      new HostImpl(info_, info_->name() + address->asString(), address, false, 1, ""));
+  dispatcher_.post([this, host]() -> void { addHost(host); });
```

Comment:
I need to think about this more (well, you do, but I'm helping 😉 ), but I think there is a race condition here with CDS cluster removal. I think it's possible that cluster has been removed on main thread by the time the post gets there. We can chat here or offline about potential solutions (and if there really is a problem).